### PR TITLE
[Graviton][build] CVE-2015-20107 fix

### PR DESCRIPTION
**Pull request by xncqr**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
resolving CVE-2015-20107 on Graviton Inference images by having the OS upgrade it's packages.

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [x] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [x] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/buildspec-graviton-1.12.yml [hunk 1]
```diff
@@ -1,9 +1,8 @@
 account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
 region: &REGION <set-$REGION-in-environment>
 framework: &FRAMEWORK pytorch
-version: &VERSION 1.10.0
```
Comment (arjkesh) [inline pytorch/buildspec-graviton-1.12.yml:4]: Please revert these buildspec changes. We track the "latest" framework version in the un-versioned buildspecs, and older framework versions in versioned buildspecs.

If you would like to build older graviton images in PR context, please set graviton_mode to True in the developer toml, and also fill in the path to the appropriate buildspec in the buildspec_overrides section (https://github.com/aws/deep-learning-containers/blob/master/dlc_developer_config.toml#L95-L96)

File: data/ignore_ids_safety_scan.json [hunk 2]
```diff
@@ -411,6 +413,7 @@
             "py2": {
             },
             "py3": {
+                "51159":"in order to to get approved version of cryptography (>=39.0.0), the python version needs to be 3.9 or above."
```
[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:416]: That is a good point. course, with 23.9 or above, the version "should" already be updated, but if it didn't we would miss it with this.

File: pytorch/buildspec-graviton-1.12.yml (hunk 1 above)
Comment (xncqr) [reply pytorch/buildspec-graviton-1.12.yml]: reverted and properly done. Thank you.

File: data/ignore_ids_safety_scan.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:417]: Note: this is not something we will solving on this PR. but will discuss offline on approaches.

[the remark this answers is not in the document]
Comment (arjkesh) [reply data/ignore_ids_safety_scan.json:417]: @xncqr I noticed here https://pypi.org/project/cryptography/#history there is no version 39.0.0 - is this the right message (related to python 3.9), or is it simply the fact that there is no "approved" version of cryptography package? Can you link the information that led you to add this to the ignore list?

[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:417]: please looks up.. it's supposed to be py3.9 and the comment was updated this morning.

[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:417]: In terms of the 39.0.0.. that is correct as released to the CVE. in that context 39.0.0 is referenced to py3.9 with subversion. So that is correct.

[the remark this answers is not in the document]
Comment (arjkesh) [reply data/ignore_ids_safety_scan.json:417]: Can you please link the CVE?

[the remark this answers is not in the document]
Comment (shantanutrip) [reply data/ignore_ids_safety_scan.json:417]: @arjkesh the skip is added to all the images based on the `python_version` defined in the buildspec file, for eg., https://github.com/aws/deep-learning-containers/blob/master/pytorch/buildspec.yml#L54.

In case we define the `python_version` for Py39 images as py3, it will apply on the Py39 images, otherwise not.

[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:417]: @shantanutrip question. can we be more specific in the json file to state the next level version? aka py37 over py3?

File: tensorflow/inference/docker/2.7/py3/Dockerfile.graviton.ec2.cpu [hunk 3]
```diff
@@ -68,7 +69,7 @@ RUN ${PIP} install --no-cache-dir \
     gevent==21.1.1 \
     requests==2.25.1 \
     grpcio==1.34.1 \
-    protobuf==3.15.0 \
+    protobuf==3.18.3 \
```
Comment (shantanutrip) [inline tensorflow/inference/docker/2.7/py3/Dockerfile.graviton.ec2.cpu:72]: Can we use this: https://github.com/aws/deep-learning-containers/pull/2243/files#diff-446e61b61a3121f41f2ea280cd7c8a966e0fa8b151133b3c5ab5fc165b42b260R162

File: data/ignore_ids_safety_scan.json [hunk 4]
```diff
@@ -228,7 +229,8 @@
                 "43453":"mxnet 1.9 currently does not work with numpy 1.20.0 or above",
                 "44715":"mxnet 1.9 currently does not work with numpy 1.20.0 or above",
                 "44716":"mxnet 1.9 currently does not work with numpy 1.20.0 or above",
-                "44717":"mxnet 1.9 currently does not work with numpy 1.20.0 or above"
+                "44717":"mxnet 1.9 currently does not work with numpy 1.20.0 or above",
+                "51159":"in order to to get approved version of cryptography (>=39.0.0), the python version needs to be 3.9 or above."
```
[the remark this answers is not in the document]
Comment (xncqr) [reply data/ignore_ids_safety_scan.json]: this looks out of date as 51159 was removed from MXnet as not needed.

File: tensorflow/inference/docker/2.7/py3/Dockerfile.graviton.ec2.cpu (hunk 3 above)
Comment (xncqr) [reply tensorflow/inference/docker/2.7/py3/Dockerfile.graviton.ec2.cpu]: Updated it to use protobuf>=4.21.6, being the latest.

File: mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu [hunk 5]
```diff
@@ -20,6 +20,7 @@ ENV PYTHONDONTWRITEBYTECODE=1 \
     TEMP=/home/model-server/tmp
 
 RUN apt-get update \
+ && apt-get -y upgrade \
```
Comment (shantanutrip) [inline mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu:23]: Lets remove this completely? It will be better to let this be handled in the future. In case we make this change today but do not release, it might become complicated to trace any bug or issue in the future.

File: data/ignore_ids_safety_scan.json [hunk 6]
```diff
@@ -412,6 +412,7 @@
             "py2": {
             },
             "py3": {
+                "51159":"cryptography>38.0.1 does not exist yet"
```
Comment (shantanutrip) [inline data/ignore_ids_safety_scan.json:415]: Since this is the only vulnerability present in `inference` for `Pytorch` lets keep it as it is. In the future, we can make a key for `inference_graviton`

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu [hunk 7]
```diff
@@ -102,6 +103,7 @@ RUN conda install -c conda-forge -y \
  && pip install --upgrade pip --trusted-host pypi.org --trusted-host files.pythonhosted.org \
  && ln -s /opt/conda/bin/pip /usr/local/bin/pip3 \
  && pip install packaging==20.4 \
+    protobuf>=4.21.6 \
```
Comment (shantanutrip) [inline pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu:106]: Not sure if this would work - there might be some dependency conflicts. You might want to fix the version as shown in https://github.com/aws/deep-learning-containers/pull/2243/files#diff-ec22ab4008f5768525a40acc503df34dd72bce7651f65e1fa87fc999548c55afR112

File: mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu (hunk 5 above)
Comment (xncqr) [reply mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu:23]: this is required for get past the sanity check for another vul-id on the OS. The current frameworks already had it.

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu (hunk 7 above)
Comment (xncqr) [reply pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu:106]: according to the testing, what is set is working as expected. 4.21.6 is compatible py3.7 and matching python apps should be working. Would rather not go backwards unless we know of an issue as the previous setting was to "be safe". the linked diff still is hard pinning down which can cause issue down the road as well, and would need to be updated. Is this what we want to do?

File: mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu (hunk 5 above)
Comment (xncqr) [reply mxnet/inference/docker/1.9/py3/Dockerfile.graviton.ec2.cpu:23]: Missed, this is MXnet.. doesn't matter here anyway. as I believe this one is deprecated

File: data/ignore_ids_safety_scan.json (hunk 6 above)
Comment (xncqr) [reply data/ignore_ids_safety_scan.json:415]: /agreed

File: pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu (hunk 7 above)
Comment (xncqr) [reply pytorch/inference/docker/1.10/py3/Dockerfile.graviton.ec2.cpu]: Made the change. Am expecting to visit this later down the road, but would likely be updating other packages at that time.

**Review by arjkesh — APPROVED** [review]
